### PR TITLE
`Add ByteBPETokenizer`

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This class wraps the ByteBPEProcessor class from Cutlery and does the pre/post-processing to provide output that is fit for transformer models. The class also provides two class methods to load a vocabulary, `from_files` and `from_hf_hub`
    
This change also adds an initial `Tokenizer` base class that all future tokenizers can implement.

I hope that all the abstractions hold up, but we can probably only know once we have more tokenizers.

### Types of change

Enhancement.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/tokenization/bbpe_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+from typing import List, Type, TypeVar
+from cutlery import ByteBPEProcessor
+import json
+from pathlib import Path
+
+from . import PiecesWithIds, Tokenizer
```
Comment (shadeMe) [inline curated_transformers/tokenization/bbpe_tokenizer.py:6]:
```suggestion
from .tokenizer import PiecesWithIds, Tokenizer
```
Potential circular import?

File: curated_transformers/tokenization/bbpe_tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            )
+
+        eos_id = self.processor.piece_id(self.eos_piece)
+        if eos_id is None:
+            raise ValueError(
+                f"Byte BPE piece encoder vocabulary doesn't contain '{self.eos_piece}' piece"
+            )
+
+        pad_id = self.processor.piece_id(self.pad_piece)
+        if eos_id is None:
```
Comment (shadeMe) [inline curated_transformers/tokenization/bbpe_tokenizer.py:84]:
```suggestion
        if pad_id is None:
```

File: curated_transformers/tokenization/bbpe_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            raise ValueError(
+                f"Byte BPE piece encoder vocabulary doesn't contain '{self.eos_piece}' piece"
```
Comment (shadeMe) [inline curated_transformers/tokenization/bbpe_tokenizer.py:86]:
```suggestion
                f"Byte BPE piece encoder vocabulary doesn't contain '{self.pad_piece}' piece"
```

File: curated_transformers/tokenization/bbpe_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            text_lens.append(1)
+
+            ids.append(text_ids)
+            pieces.append(text_pieces)
+            lens.append(text_lens)
+
+        return PiecesWithIds(pad_id=pad_id, ids=ids, lens=lens, pieces=pieces)
+
+    @classmethod
+    def _convert_hf_tokenizer(cls: Type[Self], tokenizer) -> Self:
```
Comment (shadeMe) [inline curated_transformers/tokenization/bbpe_tokenizer.py:117]:
```suggestion
    def _convert_hf_tokenizer(cls: Type[Self], tokenizer: Any) -> Self:
```
Or we go for the full annotation by importing `tokenizers`.

File: curated_transformers/tokenization/tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        for idx, seq_ids in enumerate(self.ids):
+            mask[idx, : len(seq_ids)] = True
+        return mask
+
+    @property
+    def padded_tensor(self):
+        """Padded CPU tensor of the piece identifiers."""
+        n_seqs = len(self.ids)
+        max_len = max(len(seq_ids) for seq_ids in self.ids)
+        padded = torch.full((n_seqs, max_len), self.pad_id)
```
Comment (shadeMe) [inline curated_transformers/tokenization/tokenizer.py:36]: I think this defaults to `torch.int64`. Given that 32-bits is plenty for the IDs, we might want to explicitly set `dtype=torch.int32` for better throughput on the GPU (when it eventually gets uploaded).

Comment (danieldk) [reply curated_transformers/tokenization/tokenizer.py:36]: Makes sense, I'll fix it (`int32` didn't work in the past for a lot of things like the embedding layer in the past, to I built up this habit).

Comment (danieldk) [reply curated_transformers/tokenization/tokenizer.py]: Fixed.

**Review by shadeMe — APPROVED** [review]

**Pull request merged**